enclave/core/sgx/once.c

```diff
@@ -5,34 +5,74 @@
 #include <openenclave/internal/thread.h>
 #include <openenclave/internal/utils.h>
 
+#define FUNC_NOT_INVOKED 0
+#define FUNC_BEING_INVOKED 1
+#define FUNC_INVOKED 2
+
 oe_result_t oe_once(oe_once_t* once, void (*func)(void))
 {
     if (!once)
         return OE_INVALID_PARAMETER;
 
+    oe_once_t status = *once;
```

Comment:
The optimizer will optimize **status** entirely out here and in line 29 and use the value passed on the stack. If you don't want that to happen, prefix this line with volatile.

Reply:
While it's true that volatile will prevent it being optimized out, there's a memory barrier right under the statement, so it already won't be optimized out.

Note on architectures with a weak memory model (i.e. ARM), volatile would not be sufficient, since you need to do a load-acquire., making the volatile redundant.

Reply:
@anitagov I agree with @gupta-ak

Reply:
This is being optimized out with the Compiler explorer (https://godbolt.org/) and hence I brought it up.

Reply:
@anitagov To be clear, when you say "optimized out" what do you mean?

Consider the following two functions where the value of _*once_ is read twice and added.
The second function has an _acquire barrier_ where s the first does not.

```c
#include <atomic>

int read_value(int* once)
{
    int local = *once;
    return local + *once;
}
int read_value_with_barrier(int* once)
{
    int local = *once;
    std::atomic_thread_fence(std::memory_order_acquire);
    return local + *once;
}
```
In the first function, the compiler sees that *once has already been read and reuses the value without reading the value again:

```asm
read_value(int*): 
    movl (%rdi), %eax
    addl %eax, %eax    # Reuse value in eax
    retq
```

For the second function, the _acquire-barrier_ asks the compiler to NOT move any reads after the barrier to before the barrier. Therefore the compiler emits code to re-read the value of _*once_.

```asm
read_value_with_barrier(int*): 
     movl (%rdi), %eax
     addl (%rdi), %eax     # Reread the value of *rdi i.e *once
     retq
```
You can see this example here: https://godbolt.org/z/t9VWD7

x64 is a strongly ordered acrchitecture and the compiler does not need to emit any special assembly instructions for the _acquire-barrier_.

However, x64 is an exception and special instructions are needed for ARM, Power PC, RISC and other architectures to ensure the _acquire_ semantics.

In ARM64 for example, in addition to re-reading the value of *once, the compiler also emits a [Data Memory Barrier  Instruction](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0489c/CIHGHHIE.html)

```asm
read_value(int*):
    ldr w0, [x0]
    lsl w0, w0, 1   # Reuse the value in w0
    ret

read_value_with_barrier(int*):
    ldr w1, [x0]
    dmb ishld      # Acquire barrier 
    ldr w0, [x0]    # Reread *once
    add w0, w1, w0
    ret
```

You can see the example here: https://godbolt.org/z/AK72MU

If there is still confusion, can you share the link to what you are seeing in compiler explorer?(There is a share button).

Reply:
Thanks @anakrish . I verified again with Compiler explorer that things were fine.